### PR TITLE
Codegen: Generate static dispatch headers per operator

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #69840
* #68462
* #68693
* #68692
* #68691
* #68697
* #68690
* #68689
* __->__ #68714
* #68688
* #68687
* #68247

This splits the static dispatch headers (e.g. `CPUFunctions.h`)
into per operators headers (e.g. `ops/empty_cpu_dispatch.h`) which is
needed for when `Tensor.h` is compiled with static dispatch enabled.

There are also several places in ATen where the static dispatch
headers are used as an optimization even in dynamic dispatch builds.

Differential Revision: [D32596265](https://our.internmc.facebook.com/intern/diff/D32596265)

### CODE REVIEW

File: aten/src/ATen/templates/DispatchKeyFunctions_inl.h [hunk 1]
```diff
@@ -11,14 +12,8 @@
...
 #endif
 
-namespace at {
-namespace ${dispatch_namespace} {
-
-${dispatch_namespaced_declarations}
-
-} // namespace ${dispatch_namespace}
-} // namespace at
+$DispatchKeyFunctions_inl_includes
```
Comment (albanD) [inline aten/src/ATen/templates/DispatchKeyFunctions_inl.h:19]: Missing {} ?

**Review by albanD — COMMENTED** [review]
Sound good, minor typo.

**Comment by albanD** [comment]
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32596265).

File: aten/src/ATen/templates/DispatchKeyFunctions_inl.h (hunk 1 above)
Comment (peterbell10) [reply aten/src/ATen/templates/DispatchKeyFunctions_inl.h:19]: They're actually optional:
https://github.com/pytorch/pytorch/blob/9554ebe44e6e73dc75105d4935d41e626e03299b/tools/codegen/code_template.py#L4

Comment (albanD) [reply aten/src/ATen/templates/DispatchKeyFunctions_inl.h:19]: Right, but that's the only place where we don't use them? :p 
Not very important.

**Comment by albanD** [comment]
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32596265).

**Review by albanD — APPROVED** [review]

**Comment by albanD** [comment]
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32596265).

**Comment by albanD** [comment]
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32596265).

**Pull request closed without merging**